Shared/Samples/Trace utility network/TraceUtilityNetworkView.swift

```diff
@@ -0,0 +1,539 @@
+    /// - Parameters:
+    ///   - feature: The feature to be added to the pending trace.
+    ///   - mapPoint: The location on the map where the feature was discovered. If the feature is a
+    ///   junction type, the feature's geometry will be used instead.
+    ///
+    ///   For junction features with more than one terminal, the user should be prompted to pick a
+    ///   terminal. For edge features, the fractional point along the feature's edge should be
+    ///   computed.
+    func add(_ feature: ArcGISFeature, at mapPoint: Point) {
+        if let element = network?.makeElement(arcGISFeature: feature),
+           let geometry = feature.geometry,
+           let table = feature.table as? ArcGISFeatureTable,
+           let networkSource = network?.definition?.networkSource(named: table.tableName) {
+            switch networkSource.kind {
+            case .junction:
+                add(element, at: geometry)
+                if element.assetType.terminalConfiguration?.terminals.count ?? .zero > 1 {
+                    terminalSelectionIsOpen.toggle()
+                }
+            case .edge:
+                if let line = GeometryEngine.makeGeometry(from: geometry, z: nil) as? Polyline {
+                    element.fractionAlongEdge = GeometryEngine.polyline(
+                        line,
+                        fractionalLengthClosestTo: mapPoint,
+                        tolerance: -1
+                    )
+                    updateHint(String(format: "fractionAlongEdge: %.3f", element.fractionAlongEdge))
+                    add(element, at: mapPoint)
+                }
+            }
+        }
+    }
+    
+    /// Identifies the first discoverable feature at the provided screen point.
+    /// - Parameters:
+    ///   - screenPoint: The location on the screen where the identify operation is desired.
+    ///   - proxy: The map view proxy to perform the identify operation with.
+    /// - Returns: The first discoverable feature or `nil` if none were identified.
+    func identifyFeatureAt(_ screenPoint: CGPoint, with proxy: MapViewProxy) async -> ArcGISFeature? {
+        guard let feature = try? await proxy.identifyLayers(
+            screenPoint: screenPoint,
+            tolerance: 10
+        ).first?.geoElements.first as? ArcGISFeature else {
+            return nil
+        }
+        return feature
+    }
+    
+    /// Resets all of the important stateful values for when a trace is cancelled or completed.
+    func reset() {
+        map.operationalLayers.forEach { ($0 as? FeatureLayer)?.clearSelection() }
+        points.removeAllGraphics()
+        pendingTrace = nil
+        tracingActivity = .none
+    }
+    
+    /// Performs important tasks including adding credentials, loading and adding operational layers.
+    func setup() async {
+        try? await ArcGISEnvironment.authenticationManager.arcGISCredentialStore.add(.publicSample)
+        try? await network?.load()
```

Comment:
Good call, done

<img width="635" alt="Screenshot 2023-03-24 at 18 11 27" src="https://user-images.githubusercontent.com/16397058/227674680-f239a649-ab85-4dd5-97db-2839bd963fca.png">

Reply:
It's rare that all of us missed this while reviewing this PR, but there is a case that the network itself is `nil` because the map fails to load, thus it doesn't throw an error while being in an invalid state.

It led to the issue in `swift/issues/4401`